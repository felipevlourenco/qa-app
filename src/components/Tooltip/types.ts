import { ReactNode } from 'react'

export interface TooltipProp {
  children: ReactNode
  content: string
}
