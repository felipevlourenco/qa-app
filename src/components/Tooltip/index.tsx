import { TooltipProp } from 'components/Tooltip/types'
import useToggle from 'hooks/useToggle'
import styles from 'components/Tooltip/styles.module.css'
import { useMemo, useRef } from 'react'

const Tooltip = ({ children, content }: TooltipProp) => {
  const { isToggled, toggleOn, toggleOff } = useToggle()
  const childRef = useRef<HTMLDivElement>(null)

  const offsetX = useMemo(() => {
    return childRef.current ? childRef.current.offsetWidth / 3 : 0
  }, [childRef.current])

  const offsetY = useMemo(() => {
    return childRef.current ? childRef.current.offsetHeight * 2 : 0
  }, [childRef.current])

  return (
    <div className={styles.tooltip} onMouseEnter={toggleOn} onMouseLeave={toggleOff}>
      <div className={styles.child} ref={childRef}>
        {children}
      </div>
      {isToggled && (
        <span className={styles.content} style={{ left: offsetX, top: offsetY }}>
          {content}
        </span>
      )}
    </div>
  )
}

export default Tooltip
