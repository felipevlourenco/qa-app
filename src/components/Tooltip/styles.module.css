.tooltip {
  position: relative;
  width: max-content;
}

.child:hover {
  cursor: help;
}

.content {
  position: absolute;
  top: -1.5rem;
  border: 1px solid #cccccc;
  border-radius: 4px;
  max-width: 300px;
  min-width: 300px;
  background-color: #ffffff;
  padding: 1rem;
  box-shadow: 5px 5px 5px #00000020;
}

.content::before {
  content: "\A";
  border-style: solid;
  border-width: 2px 15px 15px 15px;
  border-color: transparent transparent #ffffff transparent;
  position: absolute;
  left: 12px;
  top: -15px;
  z-index: 2;
}

.content::after {
  content: "\A";
  border-style: solid;
  border-width: 0px 15px 15px 15px;
  border-color: transparent transparent #cccccc transparent;
  position: absolute;
  left: 12px;
  top: -15px;
  z-index: 1;
}