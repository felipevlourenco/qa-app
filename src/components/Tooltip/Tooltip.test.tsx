import { fireEvent, render, screen } from '@testing-library/react'
import Tooltip from 'components/Tooltip'

describe('components/Tooltip', () => {
  const childText = 'Hover to show tooltip'
  const tooltipContent = 'This is a tooltip'

  it('should render', () => {
    const { container } = render(<Tooltip content={tooltipContent}>{childText}</Tooltip>)
    expect(container).toMatchSnapshot()
  })

  it('should be visible on mouse hover', () => {
    const { getByText } = render(<Tooltip content={tooltipContent}>{childText}</Tooltip>)
    fireEvent.mouseEnter(getByText(childText))
    expect(getByText(tooltipContent)).toBeInTheDocument()
  })

  it('should not be visible on mouse hover', () => {
    const { getByText } = render(<Tooltip content={tooltipContent}>{childText}</Tooltip>)
    fireEvent.mouseEnter(getByText(childText))
    expect(getByText(tooltipContent)).toBeInTheDocument()
    fireEvent.mouseLeave(getByText(childText))
    expect(screen.queryByText(tooltipContent)).not.toBeInTheDocument()
  })
})
