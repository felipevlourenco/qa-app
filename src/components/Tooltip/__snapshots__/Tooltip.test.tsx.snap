// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Tooltip should render 1`] = `
<div>
  <div
    class="tooltip"
  >
    <div
      class="child"
    >
      Hover to show tooltip
    </div>
  </div>
</div>
`;
